pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/StoolUploadRoute.java

```diff
@@ -0,0 +1,180 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import spark.QueryParamsMap;
+import spark.Request;
+import spark.Response;
+
+
+public class StoolUploadRoute extends RequestHandler {
+    private static final Logger logger = LoggerFactory.getLogger(StoolUploadRoute.class);
+
+
+    private static final String PARTICIPANT_ID = "participantId";
+    private static final String MF_BARCODE = "mfBarcode";
+    private static final String RECEIVE_DATE = "receiveDate";
+
+    @Override
+    protected Object processRequest(Request request, Response response, String userId) throws Exception {
+        QueryParamsMap queryParams = request.queryMap();
+        String realm;
+
+        if (queryParams.value(RoutePath.REALM) != null) {
+            realm = queryParams.get(RoutePath.REALM).value();
+        } else {
+            throw new RuntimeException("No realm query param was sent");
+        }
+        String userIdRequest = UserUtil.getUserId(request);
+
+        if (UserUtil.checkUserAccess(realm, userId, "kit_upload", userIdRequest)) {
+            String content = request.body();
+
+
+            try {
+                List<StoolUploadObject> stoolUploadContent;
+
+                try {
+                    stoolUploadContent = isFileValid(content);
+                } catch (Exception e) {
+                    response.status(500);
+                    return e.getMessage();
+                }
+
+                final List<StoolUploadObject> stoolUploadObjects = stoolUploadContent;
+                if (stoolUploadObjects.isEmpty()) {
+                    return "Text file was empty or couldn't be parsed to the agreed format";
+                }
+
+                logger.info("Text file was uploaded and parsed successfully");
+
+                StoolUploadDao stoolUploadDao = new StoolUploadDao();
+                stoolUploadObjects.forEach(stoolUploadObject -> {
+
+                    String participantId = stoolUploadObject.getParticipantId();
+                    String mfBarcode = stoolUploadObject.getMfBarcode();
+                    String receiveDate = stoolUploadObject.getReceiveDate();
+
+                    Optional<StoolUploadDto> stoolUploadDto = stoolUploadDao.getStoolUploadDto(participantId, mfBarcode);
+
+                    if (stoolUploadDto.isPresent()) {
+                        logger.info("Successfully generated kits, trying to update the table with provided parameters...");
+                        stoolUploadDao.updateKitData(receiveDate, stoolUploadDto.get().getKitId(), mfBarcode);
+                    } else {
+                        logger.warn("Unable to fetch kits with provided parameters");
+                    }
+                });
+            } catch (Exception e) {
+                return e.getMessage();
+            }
+        } else {
+            response.status(500);
+            return (UserErrorMessages.NO_RIGHTS);
+        }
+        return null;
+    }
```

Comment:
Once again, too complex logics in `Route` class, main purpose of `Route` class it to extract necessary data from the request and pass it to some so called use case classes, which are responsible to hold all complex logics and encapsulate them, for the reference you can get `CreateCohortTagRoute`, moreover I remember that the similar logic is used somewhere already probably for kit upload, instead of duplicating the same code you may introduce some abstraction which will gather the same logic for both of them.